### PR TITLE
Extensible grammar and runtime

### PR DESCRIPTION
**Pull request by jfallows**
Note: this requires a backwards incompatible change in the K3PO grammar.
Note: upgrade script provided in `tools/upgrade.sed`

Added an extensible `TypeSystem` used by the parser to support custom language elements not present in the core, such as `http:method`, `http:version` `http:header`, and `http:status`.
```
accept "http://localhost:8080/path?query"

accepted
connected

read http:method "GET"
read http:version "HTTP/1.1"
read http:header "Host" "localhost:8080"
read http:header "Upgrade" /(?i:websocket)/
read http:header "Connection" /(?i:Upgrade)/
read http:header "Sec-WebSocket-Key" /(?<key>[a-zA-Z0-9+\/=]{24})/
read http:header "Sec-WebSocket-Version" "13"

write http:status "101" "Switching Protocols"
write http:version "HTTP/1.1"
write http:header "Upgrade" "websocket"
write http:header "Connection" "Upgrade"
write http:header "Sec-WebSocket-Accept" ${ws:handshakeHash(key)}
write flush

# connection established
```

Added an extensible `BehaviorSystem` used by driver to support runtime enforcement of custom language elements.

New transport schemes can now be implemented outside the K3PO project and be discovered via `ServiceLoader` by placing the artifact on the `<dependency>` classpath of the `k3po-maven-plugin`.

### CODE REVIEW

File: driver/src/test/resources/org/kaazing/k3po/driver/internal/file/read.and.write.simultaneously.in.file.rpt [hunk 1]
```diff
@@ -15,7 +15,7 @@
 #
 
 #
-# read and write track file position independently
+#read andwrite track file position independently
```
Comment (dpwspoon) [inline driver/src/test/resources/org/kaazing/k3po/driver/internal/file/read.and.write.simultaneously.in.file.rpt]: format clean up shouldn't change this

File: driver/src/test/resources/org/kaazing/k3po/driver/internal/http/http.client.post.chunking.with.trailer.and.extensions/request.rpt [hunk 2]
```diff
@@ -26,7 +26,7 @@ write http:header "Trailer" "Trailing-Header"
 write "Chunk A"
 write "Chunk B"
 # Not supported in netty 3.9
-# write option chunkExtension "some-extension-value"
+#write option chunkExtension "some-extension-value"
```
Comment (dpwspoon) [inline driver/src/test/resources/org/kaazing/k3po/driver/internal/http/http.client.post.chunking.with.trailer.and.extensions/request.rpt]: format clean up should not change this comment

**Comment by dpwspoon** [comment]
It should be possible now to write:

```
write http:version "HTTP/1.1"
write http:header "Upgrade" "websocket"
write http:header "Connection" "Upgrade"
```

as
```
write http:version "HTTP/1.1"
      http:header "Upgrade" "websocket"
      http:header "Connection" "Upgrade"
```

**Comment by dpwspoon** [comment]
StructuredTypeInfo needs to 
```
@Override 
public String toString()
```
as it is called in several Ast describe nodes

**Comment by dpwspoon** [comment]
Should BehaviorSystem and others (WriteOptionFactory, ReadOptionFactory) be in public API (i.e. not internal packaging).  Maybe file a follow-up ticket?

**Comment by cmebarrow** [comment]
@jfallows Since connect and accept uri is now quoted will it allow syntax like 
```
connect "nukleus://streams/{$target}"
```
that we agreed we'll need for http spec tests?

**Comment by jfallows** [comment]
The `.toString()` is now added, the `upgrade.sed` script now skips "tidying" of comments and has been re-applied to the specification scripts.

The grammar updates for consecutive `write http:header` (or similar) and `"http://localhost:8080/${path}"` syntax are both enhancements beyond the scope of this PR which aims to make K3PO extensible.

**Comment by jfallows** [comment]
Investigating build breakage on Travis.

**Comment by jfallows** [comment]
Travis build working again.

**Comment by jfallows** [comment]
Note: we might need some further (minor) changes on stream validation to support custom extensions after `write close`.

**Comment by jfallows** [comment]
Note: no need to hold for any further changes.

**Pull request merged**